packages/flutter/lib/src/material/text_selection.dart

```diff
@@ -20,64 +23,406 @@ const double _kHandleSize = 22.0;
+                    });
+                  },
+                  tooltip: _overflowOpen ? 'Back' : 'More',
+                ),
+              ),
+              ...items,
+            ],
+          ),
+        ),
+      ),
+    );
+
+    // When the menu shrinks as the overflow is opened, then it occupies the
+    // same size as it did with overflow closed and aligns itself to the right
+    // of its original size.
+    return _maxWidth == null ? child : SizedBox(
+      width: _maxWidth,
+      child: Align(
+        alignment: Alignment.topRight,
+        heightFactor: 1.0,
+        widthFactor: 1.0,
+        child: child,
       ),
     );
+
   }
 }
 
-/// Centers the toolbar around the given position, ensuring that it remains on
+// Renders the menu items in the correct positions in the menu and its overflow
+// submenu based on calculating which item would first overflow.
+class _TextSelectionToolbarItems extends MultiChildRenderObjectWidget {
+  _TextSelectionToolbarItems({
+    Key key,
+    @required List<Widget> children,
+    @required this.isAbove,
+    @required this.overflowOpen,
+  }) : assert(children != null),
+       assert(isAbove != null),
+       assert(overflowOpen != null),
+       super(key: key, children: children);
+
+  final bool isAbove;
+  final bool overflowOpen;
+
+  @override
+  _TextSelectionToolbarItemsRenderBox createRenderObject(BuildContext context) {
+    return _TextSelectionToolbarItemsRenderBox(
+      isAbove: isAbove,
+      overflowOpen: overflowOpen,
+    );
+  }
+
+  @override
+  void updateRenderObject(BuildContext context, _TextSelectionToolbarItemsRenderBox renderObject) {
+    renderObject
+      ..isAbove = isAbove
+      ..overflowOpen = overflowOpen;
+  }
+
+  @override
+  MultiChildRenderObjectElement createElement() => MultiChildRenderObjectElement(this);
+}
+
+class _TextSelectionToolbarItemsRenderBox extends RenderBox with ContainerRenderObjectMixin<RenderBox, FlexParentData>, RenderBoxContainerDefaultsMixin<RenderBox, FlexParentData> {
+  _TextSelectionToolbarItemsRenderBox({
+    @required bool isAbove,
+    @required bool overflowOpen,
+  }) : assert(overflowOpen != null),
+       assert(isAbove != null),
+       _isAbove = isAbove,
+       _overflowOpen = overflowOpen,
+       super();
+
+  int _lastIndexThatFits = -1;
+
+  bool _isAbove;
+  bool get isAbove => _isAbove;
+  set isAbove(bool value) {
+    if (value == isAbove) {
+      return;
+    }
+    _isAbove = value;
+    markNeedsLayout();
+  }
+
+  bool _overflowOpen;
+  bool get overflowOpen => _overflowOpen;
+  set overflowOpen(bool value) {
+    if (value == overflowOpen) {
+      return;
+    }
+    _overflowOpen = value;
+    markNeedsLayout();
+  }
+
+  // Lay out all children, regardless of whether or not they will be painted or
+  // placed with an offset. Find which child overflows, if any.
+  void _layoutChildren() {
+    int i = -1;
+    double width = 0.0;
+    visitChildren((RenderObject renderObjectChild) {
```

Comment:
do you have to layout all the children? Can you layout the child one by one and stop when the width overflows? Also can we defer the layout of navigation bar until we know it will overflow?

Reply:
In the case where the overflow menu is open, I have to layout all of the children, because I have to decide where it overflows and then I have to paint the overflowing children.  In the case where it's not open, I can add an optimization to not layout the overflowing children.